### ISSUE TITLE
Feature request: managed dataset catalog and SampleRef Datum protocol

### ISSUE THREAD

**Issue by void-main** [nex-agi/weaver#120]
## Summary

Extend the Weaver Python SDK with a managed-dataset catalog and a backward-compatible `SampleRef` variant of `Datum`. Users discover authorized datasets, explicitly choose `name + version`, pack by model-bound sample length, and submit references without learning any Server-private binding, profile, template, tokenizer, or storage locator.

Server tracking issue: https://github.com/china-qijizhifeng/weaver-server/issues/862

Each dataset exposes an immutable `content_visibility`:

- `protected`: content and real token IDs remain hidden; the SDK exposes only length, reference identity, strict redacted results, and safe aggregate metrics.
- `public`: content is visible to an authorized caller; ordinary outputs, sampling/logprobs, and canonical JSONL download are allowed.

Visibility is separate from access policy. A public dataset is not anonymous: Server must still authorize every operation.

### Catalog

Provide sync/async equivalents of:

```python
datasets = client.datasets.list(status="published")
dataset = client.datasets.get(name="hq-math", version="2026-08")
```

Safe metadata includes `name`, `version`, `description`, `sample_count`, `recommended_ratio`, compatibility/status fields, and `content_visibility`. It excludes internal IDs, bindings, storage paths, raw schema/previews, tokenizer/template, and private profile identity.

### Datum SampleRef

Keep the existing `Datum` abstraction:

```python
datum = Datum.from_sample_ref(
    dataset="hq-math",
    version="2026-08",
    sample_idx=123,
    datum_id="occurrence-123",
)
```

The SampleRef identity is exactly `dataset + version + sample_idx`. A minimal protected training Datum is:

```json
{
  "kind": "sample_ref",
  "datum_id": "occurrence-123",
  "dataset": "hq-math",
  "version": "2026-08",
  "sample_idx": 123
}
```

Ordinary payloads without `kind` retain legacy token-in semantics. The SDK validates non-empty names/versions and a non-negative integer index. A caller cannot add visibility, `binding_id`, profile, model input, target tokens, or any storage field. Public SampleRef Datums may additionally use the ordinary Datum `loss_fn_inputs`/metadata fields; protected Datums require both to be empty.

Public-only sampling/logprob prompt shape is exactly the same reference without `datum_id`:

```json
{"kind":"sample_ref","dataset":"hq-math","version":"2026-08","sample_idx":123}
```

Server resolves this to the complete model-bound training `model_input`; there is no client-selectable chat template or “strip final assistant turn” mode.

### Lengths and packing

Add model-bound batched length resolution. Each ordered result returns only the effective `input_token_count` and an opaque equality-only `model_data_revision`. Sync/async pagination must reject mixed revisions. NexRL treats managed samples as indivisible, so the full length is sufficient; no per-chunk token counts or tokens are required.

### Results

Every managed Datum result uses `kind=sample_ref_output` and echoes `datum_id`, `sample_ref`, `input_token_count`, and Server-authoritative `content_visibility`.

Define:

```python
WEAVER_REDACTED_TOKEN_ID = -8
```

- Protected token identity arrays contain only `-8`, with exactly the true returned length. Text/messages, logits, token identities, and label-dependent per-token values are unavailable.
- Public managed results use ordinary real token/output values.
- Result helpers join by `datum_id`, require an exact one-to-one ID set, and verify wire order/count/length invariants. Repeated occurrences of the same dataset row remain distinct.
- `-8` is legal only in an explicitly protected managed output and can never be converted into a token-input Datum.

### Public download

Add sync/async download of the canonical normalized messages/tools JSONL for authorized public datasets. Stream to a `.part` file, require a positive bounded `Content-Length` and valid SHA-256, verify both while streaming, `fsync`, atomically publish, and refuse overwrite by default. Clean up partial files on all failures. The SDK must never receive a GPFS path.

## Operation matrix

| Operation/capability | protected | public |
|---|---:|---:|
| model-bound lengths | yes | yes |
| built-in CE `forward_backward` | yes | yes |
| SampleRef loss inputs/metadata | must be empty | ordinary supported behavior |
| non-empty loss config/custom/surrogate | no | ordinary supported behavior |
| binary tensor transport | no; default JSON only | yes |
| `sample` / `compute_logprobs` | no | yes |
| real token identities / label-dependent per-token output | no | yes |
| canonical JSONL download | no | yes, authorized |

Protected preflight may avoid an obviously invalid call using catalog metadata, but Server remains authoritative and must reject races/spoofing. Missing or unknown visibility is treated as protected.

Exact target logprobs and elementwise CE loss are prohibited for protected data because a caller with the same model can enumerate ordinary token-in candidates and match their values, recovering the hidden token one position at a time. Equal-length `-8` alone does not close this oracle.

## Acceptance criteria

- Catalog list/get and public download have sync/async parity, pagination/error handling, and no private fields.
- `Datum.from_sample_ref` serializes through existing forward/backward helpers without changing legacy payloads.
- Protected calls fail locally where determinable and fail closed at Server for every unsupported config/operation/output; public calls retain ordinary behavior.
- Length parsing, `model_data_revision`, result alignment, strict `-8` validation, sampling prompt serialization, download atomicity/integrity, and legacy compatibility have unit tests.
- Documentation states the residual memorization/extraction risk of allowing users to obtain a model trained on protected records.

**Comment by void-main** [comment]
## Design update: model-bound per-sample length lookup

Client-side token-budget batching does not need managed token IDs or segment metadata. Because NexRL preserves one logical Datum per sample, the SDK only needs to expose one `input_token_count` for each authorized `SampleRef`.

Add a batched length-resolution capability scoped to an already-created model/run. The exact method name can follow existing SDK conventions, but the response contract should be equivalent to:

```python
SampleRefLength(
    sample_ref=SampleRef(
        dataset="hq-math",
        version="2026-08",
        sample_idx=123,
    ),
    input_token_count=1535,
)
```

Contract:

- The caller supplies public dataset name, version, and sample index plus the existing model/run identity.
- The caller does not supply or receive a tokenizer, chat template, internal materialization profile, storage binding, token IDs, loss mask, messages, or segment lengths.
- `input_token_count` is the final materialized Datum `input_tokens` length: model-bound rendering/tokenization, effective max-length truncation, then autoregressive shift, so it corresponds to `L - 1`.
- It excludes Trainer padding, THD packing alignment, and context-parallel padding.
- Support batched lookup and preserve request order or return an explicit correlation key so NexRL can resolve lengths efficiently and safely.
- The returned value is safe batching metadata only and must never be converted into placeholder token arrays on the client.

This is separate from operation responses that redact token-bearing arrays with `-8`. For batching, returning the integer length alone is sufficient.

Additional acceptance criteria:

- Sync and async APIs have parity.
- Duplicate occurrences of the same reference can be correlated independently when needed, while their resolved length is identical under the same model/run policy.
- Length lookup propagates authorization, incompatibility, missing-version, out-of-range, and materialization-policy errors without exposing protected details.

**Comment by void-main** [comment]
## Implementation and validation update\n\nImplementation branch: codex/managed-datasets-sdk\nCurrent head: e95258f\n\nImplemented:\n- managed dataset list/get catalog APIs with sync/async parity;\n- SampleRef as a Datum variant, including attached safe loss inputs;\n- batched model-bound input_token_count resolution;\n- datum_id based result alignment for multi-call flows;\n- sample operation rejection for SampleRef;\n- managed output parsing with response-only -8 placeholders;\n- fail-closed token identity, text, logits, malformed shape, and non-finite output checks;\n- bounded retries and explicit readiness error handling.\n\nThe package version is now 1.17.0. NexRL declares nex-weaver 1.17.0 or newer.\n\nVerification:\n- 604 SDK unit tests passed with integration tests excluded;\n- 29 focused managed-dataset tests plus lint, Pylint, and mypy passed;\n- the formal 30-sample cross-component smoke test confirmed that clients require only lengths and references for batching and never receive managed token IDs.\n\nRelease dependency: SDK 1.17.0 is the first artifact in the cross-repository release order.

**Comment by void-main** [comment]
Implementation PR opened: https://github.com/nex-agi/weaver/pull/121\n\nCurrent head: e95258f. Review the PR description for the public wire contract, redaction policy, test evidence, and the SDK-first release dependency.

**Comment by void-main** [comment]
PR 121 follow-up head: 6003f3d. Cross-repo review added shared batch normalization so mixed SampleRef/token batches and partially identified token batches always carry unique occurrence IDs, while all-unidentified token-only wire remains unchanged. Output alignment and loss-input reattachment use the same deterministic mapping.

**Comment by void-main** [comment]
Phase-one scope clarification (supersedes the public sampling/download/custom-input sections above):

- Keep catalog list/get, SampleRef as a Datum variant, model-bound length lookup, datum_id correlation, model_data_revision, and protected/public inline output parsing.
- Every SampleRef, public or protected, is limited to built-in cross_entropy forward_backward with empty loss config, loss inputs, per-datum metadata, and default JSON transport.
- SampleRef forward, custom/surrogate loss, sample, compute_logprobs, output reattachment, binary transport, and dataset download are not exposed.
- Public may preserve real inline values; protected accepts only length-preserving -8 token identity and no label-dependent per-token outputs.

Implementation: https://github.com/nex-agi/weaver/pull/121 at 577528e. Full SDK suite: 627 passed.

**Comment by void-main** [comment]
SDK 后续配套异步 ingestion API：在 developer/operator client 中为 dataset version 提供 `submit_source(...) -> DatasetIngestionJob`，参数使用 typed source descriptor（GPFS/TOS、immutable version、sha256/etag），并暴露安全的 `get_status/retry/cancel`；普通用户 client 仍只看到有权限的 dataset name/version/description/sample_count，不得看到 source locator、内部 profile、tokenizer/template 或 job 错误原文。默认输出/日志/repr 必须对 locator 和 credential 做 redaction，重试通过 Idempotency-Key 绑定同一 version/input digest。第一期 SDK 继续只消费已经 published 的 name+version 和 SampleRef。

**Comment by void-main** [comment]
Authoritative phase-one contract; this supersedes earlier comments where they mention `model_data_revision`, empty caller loss inputs, content/config/tokenizer/template SHA gates, CAS, source allowlists, or digest-based idempotency:

- Managed identity is the explicit public `dataset` + `version` + `sample_idx`; authorization and model-profile binding stay in Weaver Server.
- Dataset lifecycle/policy management is developer-only. Authorized users see catalog metadata and use references; they do not receive protected messages, token IDs, masks, profile details, or GPFS locators.
- Phase one is SFT: SampleRef supports built-in `cross_entropy` `forward_backward`. `sample`, `compute_logprobs`, forward-only, custom/surrogate losses, and RL are out of scope.
- Caller-owned loss inputs and request metadata are supported and remain aligned by `datum_id`. Server-owned `model_input`, `target_tokens`, `loss_mask`, and `weights` cannot be overridden.
- Protected output keeps true sequence length and replaces token identity with `-8`; token-derived per-token fields are not returned. Public managed output may expose real tokens, but phase-one request scope remains SFT.
- NexRL plans source quotas, repeat/sample, concat virtual indices, deterministic global shuffle, then lazy whole-datum token-budget batching. It never splits one sample and needs only shifted `input_token_count`.
- Raw canonical messages and model-bound variants live under `/gpfs/shared_data/.weaver`; variant preprocessing is server-selected and follows ms-swift chat-template/SFT loss-mask semantics.
- The later asynchronous ingestion API should accept a developer-submitted dataset address, return a job, and let an offline GPFS-mounted worker canonicalize/index/build variants. Do not add SHA/CAS/allowlist/config-hash gates unless they become a separately approved requirement.
